### DISCUSSION TITLE
Enabling QSV in ARM Docker Container

### DISCUSSION THREAD

**Discussion by DZaremba14600** [automatic-ripping-machine/automatic-ripping-machine#1120] [General]
I've worked with ARM since autumn 2023 and one thing that never seemed to work reliably was QSV support in the Docker container.  Only once did I get the QSV indicator on the ARM main page to light green and it quit quickly thereafter and I was never able to replicate that.  I used the ubuntu-quicksync.sh script on the host, from within the container, all kind of other things and nothing seemed to work again.

I decided to do a non-docker or manual or bare-metal ARM install onto a clean install of Ubuntu 22.04 and go from there.  Very quickly I had success with a QSV enabled install of ARM.  I found I didn't end up really using QSV, but it was working and it was quite the performance improvement when used.

I detailed that adventure in the post entitled 'My recent ARM experience with Ubuntu 22.04 bare (no Docker) and Handbrake custom profiles - I hope it helps somebody.'.  From the response, it seems it was helpful for some and I'm glad for that.

However, I am a fan of Docker and the path forward for the project is definitely ARM in a Docker container.  

This QSV thing was still bugging me.  I figured it cant be THAT hard to get it going, so I decided to work on it by building the Docker image from scratch and seeing what needed to be done to get it working.

I pulled the project and found that the build process had some issues.  That was reported and got quickly resolved and I got to work.  

Great Support!!

3 things need to be in place for QSV support and one was already there - Handbrake was being built with QSV support enabled.  That left getting the interface  to the GPU in place and getting the software support for QSV into place.

The interface to the GPU is provided by installing 'intel-media-va-driver'. 'intel-media-va-driver' comes in two flavors, one is open sourced (intel-media-va-driver) and the other has closed source enhancements/extensions (intel-media-va-driver-non-free).

The software support for QSV is provided by the 'Intel Media SDK'.  That can be found at 'Intel Media Stack on Ubuntu' - https://github.com/Intel-Media-SDK/MediaSDK/wiki/Intel-media-stack-on-Ubuntu.

I decided to first try loading this support into the ARM Docker container by accessing the container using Portainer and getting to a shell prompt, installing these modules, and seeing what happens.

All this being said, --

 I created a script called 'enable-qsv.sh' that runs 'apt update', installs one of the 'intel-media-va-driver' versions, and installs the 'Intel Media SDK'.  
I copied it to the ARM 'config' folder, and flagged it 'Execute'.
I got into Portainer and got to a prompt in the ARM Docker container
I typed in  './config/enable-qsv.sh' and let it run.  It only adds about 68 meg to the container, so it ran quickly.  No errors.
I exited the shell back to Portainer.

I fired up a browser, went to the ARM :8080 page and...  Success..!!  the QSV status indicator was green!

Was it really so, was there really QSV support?

I went back to the container and gained access to the prompt, and this time I typed - 'HandBrakeCLI --help | grep -A12 "Select video encoder" '.  There was a list of supported encoders and it included qsv_h264 and qsv_h265.  Great!

One thing I did discover was that Handbrake seems to require the 'non-free' driver 'intel-media-va-driver-non-free' and the additional functionality it provides.  Loading the 'free' driver and the Media SDK will light up the ARM QSV green indicator on the front browser status page (it sees some QSV support), but with that config, Handbrake does not see the enhanced QSV support, Handbrake only sees that when using the 'non-free' media drivers.  Sorry to the purists, but that's out of my control.

So now it seemed like I could get QSV support going in the ARM Docker container by running this script, but I sure didn't want to have to run a .sh script every time I need to recreate the container, so I need to build an image with these changes.

Looking the .sh script I created, there's only 3 lines, all 'apt install' type lines, that need to get into the docker build.  Since they are all related to Handbrake and could be considered a dependency to proper or complete functionality of Handbrake, I added the lines to the script file 'install_mkv_hb_deps.sh' in arm-dependencies/scripts at line 24 just before cleanup.

I built arm-dependencies, built automatic-ripping-machine and recreated the container.  Now everything was reporting QSV support without the trouble of running any scripts.

I haven't ripped any disks as a test yet, I'll be doing that this evening (it was late by the time I was satisfied with how this all ended up), but I don't anticipate any issues.  My changes seem to be fairly minor and are just additional modules.

All in all some fairly minor changes that will give QSV support for ARM in a Docker container.

As a bonus, I now fire up ARM through Portainer using a Docker-Compose script.

Sorry this was so long.

Thanks.

**Comment by DZaremba14600** [comment]
I didn't have the two files I wanted to include when I posted the original message.  Here they are:

1)  Docker-Compose file - compose.yaml

2) enable-qsv.sh - this file is self documented and is used to enable QSV into an existing ARM docker container.  

Place this file in the CONFIG folder, flag it Execute, get to a shell prompt inside the ARM docker container (I use Portainer), and type   ./config/enable-qsv.sh   and the QSV support will be installed.  no restart of the container is required. This assumes QSV is configured properly on the host.

To include QSV support during the ARM docker image build, there is documentation in the .sh file itself, but basically copy the lines:

apt install -qqy intel-media-va-driver-non-free
apt install -qqy vainfo
apt install -qqy libmfx1 libmfx-tools

to line 24 before cleanup in install_mkv_hb_deps.sh in arm-dependencies/scripts

[compose.yaml.zip](https://github.com/automatic-ripping-machine/automatic-ripping-machine/files/15169229/compose.yaml.zip)
[enable-qsv.sh.zip](https://github.com/automatic-ripping-machine/automatic-ripping-machine/files/15169230/enable-qsv.sh.zip)

Let me know if there are any questions.

I have not yet done any extensive testing, but it all seems to work.

Hope this helps somebody.

**Reply by charmarkk** [reply]
Another great guide, thanks so much! Hope you or the devs can add this to the wiki. :)

When getting to the step to add the lines to the dependencies script - assuming I followed the [docker install instructions](https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/docker), where does one _find_ that ..._deps.sh script? Just like you, I got it working by doing it manually, but I'm assuming there's somewhere else those few lines have to go even when following the install guide? I'm also using a compose file rather than command and all, if that helps. Still very new to learning docker. :)

**Reply by DZaremba14600** [reply]
charmarkk,

My sincere apologies for taking so long to get a reply back to you.  I got and e-mail that you had posted a reply to my original post, but I got it Memorial Day and was not only off for the holiday weekend and holiday, but the week that followed got me so wrapped up with some things (work and personal), that I just couldn't get to a response.  Again, my apologies.


To answer your question - in the ARM project there is a folder in the project source named arm-dependencies.  in there is a folder named scripts.  In the scripts folder are a few utility scripts, but there is a script named 'install_makemkv.sh' that installs makemkv into the docker image,  and there is a script call 'install_handbrake.sh' that installs handbrake into the docker image.  these scripts download source and build and install the packages into the docker image.  there is another script called 'install_mkv_hb_dep.sh' that installs dependencies that are required to build and run the two packages respectively.

The script you seem to be referring to is 'install_mkv_hb_dep.sh'.  

The path once pulled with git would be ./arm/arm_dependencies/scripts

However:

Ok - 

I originally performed these steps and got this working.  However, with statements in regard to this project of focusing on the Docker implementation, I figured I should get more familiar with the Docker build itself.

I went to do a clean build of the Docker image/container, and while I get no errors, I cant get QSV support working properly myself.

I can take my original fix that I documented here, but I cant duplicate a working Docker build from scratch.
I also discovered that Handbrake released a new version and it seems their dependencies changed - I had to cut and past them from their site.  And there seems to be links to two different sets of Ubuntu dependencies.  <sigh>.  But I got over that.

I can get ARM to recognize that the system has QSV (the green light on the ARM homepage), but for the life of me I cant get QSV support to appear when executing 'HandBrakeCLI --help | grep -A12 "Select video encoder"' inside the container.  None of the QSV encoders show QSV support.  Its likely missing packages/support in the Docker image I am creating, but its driving me nuts.  For some reason the :latest image sees QSV is available but Handbrake doesn't see it and my fix doesn't seem to work anymore.  

In the DOCKER file for arm-dependencies, there is an image 'phusion/baseimage:focal-1.2.0' that is the 20.04 FUSION version of Ubuntu that's used as BASE and I decided to try phusion/baseimage:jammy-1.0.4 which is 22.04 JAMMY version of Ubuntu.  I figured more up to date might be better?  That's where I left off last week.  I could build a Docker image based on JAMMY successfully, QSV seen by ARM but not by Handbrake.

I'll be working on this the next few days and will provide updates.

But I gave you some info on where to find the dependencies script.

Hope this helps.

I'm in a rush and hope this post makes sense.  I'll review it and make any clarifications later.  

I'll keep you posted.

Sorry again for the delay.

**Reply by charmarkk** [reply]
You're all good! I'm in no rush but I appreciate it anyhow. 

I do see where arm-dependencies is, and I know where the script is, bt I don't think it coems through if you do the basic Docker install described in the wiki. And the QSV steps in the wiki don't work for Docker either (since that's for the host system).

> I originally performed these steps and got this working. However, with statements in regard to this project of focusing on the Docker implementation, I figured I should get more familiar with the Docker build itself. I went to do a clean build of the Docker image/container, and while I get no errors, I cant get QSV support working properly myself.

And all that is what I don't get. I get how to pull it all, I _think_ I get how to build the image, but the instructions for it are just vague enough - and there's so much jargon around Docker - that I get confused and something doesn't work, AND I don't know what specifically in the [arm-dependencies instructions](https://github.com/automatic-ripping-machine/arm-dependencies/tree/3b7c1106a4e0cc8a3506c7cfd23e905b995a65e4) applies to what, or what order to do things in, etc.  

e.g.: 
> Docker Container
> To base your docker container on arm-dependencies, add this to the top of your Dockerfile:

I don't know how all that applies to the Docker install described in the wiki. Conceptually I _think_ I do, but practically I am just not getting it. Especially when using a compose.yml file.

**Reply by DZaremba14600** [reply]
charmarkk,

Well, I finally had enough time off from this and cleared my head.  I worked on this 'QSV working in Docker' again over this past weekend and I think I got it set.  I was able to reliably build the Docker image and reliably create a Docker container that has QSV support working.  I'm doing some final testing.  I thought I had this licked a few months ago, but it ended up not being reproducible long term.  I'll explain in a future post.  And my apologies to anybody who wasted time with this, but some of the issues weren't totally my fault.  but still...

So, I'm doing some final testing and will follow this up with a longer post and hopefully provide my insights into QSV support in a Docker container.

Everything SEEMS to work right now except H.265 10-bit QSV - that encoder does no show up using the Handbrake CI status command that displays all the encoders available.  Even AV1 QSV for ARC cards that support that.  The lack of H.265 10-bit QSV might be a bug in the GPU driver I'm using according to a posting I found.  I'm checking that and I'll be trying on a different GPU/driver combination in testing.

I told you I'd get back to you - I didn't forget, and I'll post my conclusions as soon as I'm confident they can be reliably reproduced.

Thanks.

**Comment by DZaremba14600** [comment]
One last comment I forgot - 

there are devices that need to be exposed to the container from the host.  They are:

 - /dev/dri/renderD128:/dev/dri/renderD128
 - /dev/dri/card0:/dev/dri/card0

These are in the docker compose file but if you don't use compose, they will need to be included on the docker command line that starts up the container.

Basically, the flow is:  Handbrake in the ARM Docker container talks to the Intel Media SDK, which in turn talks to 'intel-media-va-driver-non-free', which in turn talks to the two /dev/dri devices which then get passed out to the corresponding devices on the host which then talk to the actual GPU (QSV).

**Comment by RedNeckGamer88** [comment]
Thank you so much. This worked for me and saved me a ton of time. Could you please update the wiki. https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/intel-qsv

**Reply by DZaremba14600** [reply]
Thanks for the response.  I'm glad it helped you.  I found it saved me a ton of time too, even with higher quality settings.

I also documented this more recently in a thread/topic entitled 'QSV Support Revisited - SUCCESS!'.  you may find something of help there also. 

I need to document hos to easily use Custom Presets.  I'll try to add that somewhere along the line.  Using those has proven useful - less crazy HB_ARGS.

I'll take a look at updating the wiki.  Will that be ok just updating that or do I need to be authorized to do so?

Thanks.

**Reply by microtechno9000** [reply]
> Thanks for the response.  I'm glad it helped you.  I found it saved me a ton of time too, even with higher quality settings.
> 
> I also documented this more recently in a thread/topic entitled 'QSV Support Revisited - SUCCESS!'.  you may find something of help there also. 
> 
> I need to document hos to easily use Custom Presets.  I'll try to add that somewhere along the line.  Using those has proven useful - less crazy HB_ARGS.
> 
> I'll take a look at updating the wiki.  Will that be ok just updating that or do I need to be authorized to do so?
> 
> Thanks.

Thanks for helping users

To update the Wiki, fork ARM, make changes on your fork and then raise a Pull Request.
See the wiki for help https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/Contribute-Wiki

**Reply by DZaremba14600** [reply]
Thanks.  I'll take a look.

**Comment by DuaneAbrames** [comment]
I am currently chasing this holy grail... 

I had a complication in just getting QSV to show as available outside the docker.  I am running on a mini-pc with an Intel N150 CPU.  This is a really new chip at this point, and what I found online said support was added in Linux 6.11.  In order to get that kernel, I am running the box on the latest Ubuntu (24.10).  That worked fine, but the docker is confused.  I get the green check, but HandbrakeCLI does NOT show any qsv encoders.  I realized that the docker build is based on jammy, and I am suspecting that since I needed Oracular Ocelot on the host system to get it working, I may need that in my docker too, so I am working through getting the dependency docker to build now,  So far I have had to fix some lines which were using something called install_clean as a wrapper for apt (replaced with apt install -qqy) and a couple of dependency issues (curl was missing) and python / pip complaining about system-installed packages.  I will write more once I get through, or give up.

**Reply by DuaneAbrames** [reply]
Replying to myself to add....

I have come to the conclusion that my issues are all driver / media foundation version issues.  I have thus far been unable to get a working container based on a newer OS, likely due to my dearth of knowledge about building dockers from scratch.  I have discovered that the HandbrakeCLI in the official Oracular Ocelot package has qsv baked in.  In the host system (non-docker) I didn't have to compile anything, simply adding the packages mentioned above got me a working vainfo (which I could not get at all under 20.04) and a Handbrake that indicates qsv codecs are available.  I am going to see if I can get ARM installed bare-metal today.  It's (as OP said) not the preferred method, but until a base-image becomes available where newer drivers / media foundation can be installed, it may be my easiest way to a fully working ARM with QSV.  This processor is NOT FAST, so I am hoping that with QSV I will be able to get Handbrake going faster than its current average of about 35fps.

**Reply by DuaneAbrames** [reply]
Getting a bare-metal install working was pretty easy, compared to the arcane rabbit-hole that was attempting to roll my own docker based on ubuntu 24.10. I needed to do this because I am running ARM on the new Intel N150 CPU.  Here are some notes for anyone who needs to do it.  I ran the commands from the official "ubuntu-20.04-install.sh" manually one at a time, with the changes mentioned below:

In the "ubuntu-20.04-install.sh":
>omit line 63
 at line 76 replace "python-psutil" with "python3-psutil"
 omit lines 137-144 (we don't need docker to run locally)
 before line 259, (or before the commands in install_armui_service) run the below commands to install python dependencies:
	apt install python3-bcrypt python3-netifaces python3-flask python3-flask-sqlalchemy python3-flask-migrate python3-flask-cors python3-flaskext.wtf python3-flask-login python3-prettytable python3-musicbrainzngs python3-xmltodict python3-waitress 
	pip install apprise --break-system-packages
	pip install discid --break-system-packages
	pip install pydvdid --break-system-packages
 
( someone with more python knowledge than I might know how to do the last 3 without the scary switch, but this worked on my system, and seems not to have broken anything else. )

after processing that scriopt, you may need to:

> sudo usermod -a -G video arm
sudo usermod -a -G render arm`

**Reply by DuaneAbrames** [reply]
I now have ARM running on bare metal, and will do some testing.  As of this writing, I have got it up and running, but have not ripped any discs yet.  Will update if anything else needs to be tweaked.

**Reply by SylvainMT** [reply]
> I now have ARM running on bare metal, and will do some testing. As of this writing, I have got it up and running, but have not ripped any discs yet. Will update if anything else needs to be tweaked.

QSV on Baremetal is a breeze I have been using that for almost a year now (I took a long time developing my script).   Assuming you are using the Debian install.  Just install the correct Intel Driver you can figure out which one you need to install here: [Debian wiki Hardware Video Acceleration page](https://wiki.debian.org/HardwareVideoAcceleration) and voila.   The BareMetal install script also builds HandBrake with the necessary libraries for AMD and NVidia.  While I haven't tested those yet, it should be as simple as installing the correct driver for your video card.

**Comment by DZaremba14600** [comment]
Sorry for the delay getting back to you.

 

I’m assuming that you’re trying to get QSV support working in a Docker Container?  Is that true?  Are you trying to build the Docker image ?

 

I could be mistaken, but I don’t think Ubuntu Server Edition should make much difference, HOWEVER, if you’re trying to run ARM on Ubuntu Server Edition in a Host situation, the Server Edition may not have the required media support packages installed that the Desktop Edition would perhaps have by default.

 

I googled that error and most of the hits referenced calling FFMPEG.

 

Could you be more specific about software you’re using and the versions?

 

I will say that I have QSV support running for the last several ARM versions, my modifications are in only one module used for the Docker image build.  There are about 10 packages installed for what Intel refers to as ‘Compute/Media’ support and their respective dependencies.  It’s been running really well for me.

 

I had some issue with my system, not related to any of this, and I reloaded Ubuntu and basically brought the system back up from scratch.  I’m currently running some tests before I resubmit a pull request.  My previous pull request raised some performance questions related to processing DVD’s so I’ve been doing some testing before I submit anything.

 

But if you get some specifics, maybe I can help.

 

Thanks.

 

David Z.

 

PS, I’m replying to the copy that git forwarded to me.  I cant seem to find the message you posed in the message stream on git’s site.

**Comment by aaronm120** [comment]
Piggybacking on this thread - Does the [script here](https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/intel-qsv) apply to getting QSV to work in a docker container as well? Or is it only when Handbrake is running outside of the container?

**Comment by itslauh** [comment]
Stumbled upon this thread. - 
I found [this](https://github.com/automatic-ripping-machine/automatic-ripping-machine/issues/1248#issuecomment-2520524653) dockerfile. Based upon that i made this Dockerfile & compose.yml (which is baed on [an earlier comment](https://github.com/automatic-ripping-machine/automatic-ripping-machine/discussions/1120#discussioncomment-9278611) 
```yml
# compose
services:
    arm:
        build: .
        environment:
            - ARM_UID=2006
            - ARM_GID=2006
            - TZ="UTC"
        ports:
            - 8080:8080
        volumes:
# Change where necessary
            - "/home/arm:/home/arm"
            - "/mnt/prc:/home/arm/process"
            - "/mnt/media/music:/home/arm/music"
            - "/mnt/media/logs:/home/arm/logs"
            - "/mnt/media/media:/home/arm/media"
            - "/mnt/media/config:/etc/arm/config"
        devices:
# Change where necessary
            - /dev/sr0:/dev/sr0
        privileged: true
        restart: always
        cpuset: "1,2"
    portainer:
        ports:
            - 9000:9000
        volumes:
            - "/var/run/docker.sock:/var/run/docker.sock"
            - "portainer_data:/data"
        privileged: true
        restart: always
        image: portainer/portainer-ce
volumes:
    portainer_data:
```
```Dockerfile
# Dockerfile
FROM automaticrippingmachine/automatic-ripping-machine:latest
RUN apt update -y &&\
    apt upgrade -y &&\
    apt install -y libmfx1 libmfx-tools libmfx-dev vorbis-tools wget &&\
    mkdir -p quicksync && cd quicksync &&\
    git clone https://github.com/intel/libva.git libva &&\
    cd libva &&\
    ./autogen.sh &&\
    make && make install &&\
    cd .. &&\
    wget https://github.com/Intel-Media-SDK/MediaSDK/releases/download/intel-mediasdk-21.3.5/MediaStack.tar.gz &&\
    tar -xvf MediaStack.tar.gz &&\
    cd MediaStack &&\
    bash install_media.sh &&\
    cd .. &&\
    git clone https://github.com/Intel-Media-SDK/MediaSDK msdk &&\
    cd msdk &&\
    mkdir build && cd build &&\
    cmake .. &&\
    make && make install &&\
    cd .. &&\
    rm -rf quicksync
```
This seems to work on my system (judging by the checkmark), which is on a [VM in proxmox](https://github.com/automatic-ripping-machine/automatic-ripping-machine/discussions/965).

**Reply by kpupp** [reply]
Just an FYI, the checkmark is _not necessarily_ indicative of QSV actually working in practice. I think there's other issues logged about that. I'm interested to give this a shot, but it looks like it's not as simple as writing the compose, right?

**Reply by itslauh** [reply]
I wouldn't be surprised, however, as I have an intel i5 4460 (which should be supported according to the Intel specs), which ,according to some, isn't supported anymore, so I can't really attest to that
